src/System.Diagnostics.Process/tests/ProcessTests.Unix.cs

```diff
@@ -783,6 +783,61 @@ public void TestProcessRecycledPid()
             Assert.True(foundRecycled);
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task Kill_ExitedNonChildProcess_DoesNotThrow(bool killTree)
+        {
+            // In this test, we kill a process in a way the Process instance
+            // is not aware the process has terminated when we invoke Process.Kill.
+
+            using (Process nonChildProcess = CreateNonChildProcess())
+            {
+                // Kill the process.
+                int rv = kill(nonChildProcess.Id, SIGKILL);
+                Assert.Equal(0, rv);
+
+                // Wait until the process is reaped.
+                for (int i = 0; i < 100 && rv == 0; i++)
+                {
+                    rv = kill(nonChildProcess.Id, 0);
+                    if (rv == 0)
+                    {
+                        // process still exists, wait some time.
+                        await Task.Delay(100);
+                    }
+                }
+                Assert.NotEqual(0, rv);
+
+                // Call Process.Kill.
+                nonChildProcess.Kill(killTree);
+            }
+
+            Process CreateNonChildProcess()
+            {
+                // Create a process that isn't a direct child.
+                int nonChildPid = -1;
+                RemoteInvokeHandle createNonChildProcess = RemoteInvoke(() =>
+                {
+                    using (RemoteInvokeHandle sleepProcess = RemoteInvoke(RemotelyInvokable.Sleep, RemotelyInvokable.WaitInMS.ToString(),
+                        // Don't pass our standard out to the sleepProcess or the ReadToEnd below won't return.
+                        new RemoteInvokeOptions { StartInfo = new ProcessStartInfo() { RedirectStandardOutput = true }
+                        }))
+                    {
+                        Console.WriteLine(sleepProcess.Process.Id);
+
+                        // Don't wait for the sleepProcess to exit.
+                        sleepProcess.Process = null;
+                    }
+                }, new RemoteInvokeOptions { StartInfo = new ProcessStartInfo() { RedirectStandardOutput = true } });
+                using (createNonChildProcess)
+                {
+                    nonChildPid = int.Parse(createNonChildProcess.Process.StandardOutput.ReadToEnd());
+                }
+                return Process.GetProcessById(nonChildPid);
```

Comment:
What would happen to the test if things got backed up in CI such that 30s passed and the sleepProcess actually completed / exited?

Reply:
The test expects the `Process.GetProcessById` and `kill SIGKILL` to happen before the `sleepProcess` exits (30 seconds currently). Otherwise it will throw.

Reply:
I'm worried we'll see this come back as sporadic failures in CI.  Can we instead make the child process wait for the parent process to exit (e.g. have the parent process pass its pid to the child and then have the child do Process.GetProcessById(pid).WaitForExit()?  If the test works correctly, the parent process should terminate the child process, and if something goes wrong and the parent process is terminated, the child process should exit.  It could also be some other scheme, I just like to avoid these timeout based tests, as they frequently lead to flakiness.

Reply:
>  Process.GetProcessById(pid).WaitForExit()

I'll add that. I'll also change the for-loop to a `while(true)`: since we successfully delivered SIGKILL, we expect the child to terminate eventually.